File: src/com/telmomenezes/synthetic/randomwalkers/RandomWalkers.java
package com.telmomenezes.synthetic.randomwalkers;

import java.util.Arrays;

import com.telmomenezes.synthetic.DiscreteDistrib;
import com.telmomenezes.synthetic.Net;
import com.telmomenezes.synthetic.Node;


public class RandomWalkers {
	private int maxLength;
	private int steps;
	
	private Net net;
	private boolean directed;
	
	private int[] dmatrix;
    private int nodes;

    private RandomWalker[] walkers;
    
    
    public RandomWalkers(Net net, boolean directed) {
    	maxLength = 5;
    	steps = 1;
    	
    	this.net = net;
    	this.directed = directed;
        
    	nodes = net.getNodeCount();
    	
        dmatrix = new int[nodes * nodes];

        init();
    }
    
    
    public void init() {
    	// clear matrices
        int largeVal = 9999999;
        //int largeVal = maxLength + 1;
        Arrays.fill(dmatrix, largeVal);
        for (int i = 0; i < nodes; i++) {
        	dmatrix[(i * nodes) + i] = 0;
        }
        
        // init walkers
        walkers = new RandomWalker[nodes];
        int i = 0;
        for (Node node : net.getNodes()) {
        	walkers[i] = new RandomWalker(node, directed, maxLength);
        	i++;
        }
    }
    
    
    public int getDist(int x, int y) {
        if (x == y) {
            return 0;
        }
        return dmatrix[(y * nodes) + x];
    }
    
    
    public void setDist(int x, int y, int d) {
    	if (getDist(x, y) <= d) {
    		return;
    	}
    	
        dmatrix[(y * nodes) + x] = d;
        
        if (!directed) {
        	dmatrix[(x * nodes) + y] = d;
        }
    }
    
    
    public void step() {
    	for (int i = 0; i < nodes; i++) {
    		RandomWalker walker = walkers[i];
    		if (walker.getOrig().getDegree() > 0) {
    			for (int j = 0; j < steps; j++) {
    				walker.step();
    				if (walker.isForward()) {
    					setDist(walker.getOrig().getId(), walker.getTarg().getId(), walker.getLength());
    				}
    				else {
    					setDist(walker.getTarg().getId(), walker.getOrig().getId(), walker.getLength());
    				}
    			}
    		}
    	}
    }
    
    
    public RandomWalkers allSteps() {
    	for (int i = 0; i < net.getEdgeCount(); i++) {
    		step();
    	}
    	
    	return this;
    }
    
    
    public void recompute() {
    	init();
    	allSteps();
    }
    
    
    public DiscreteDistrib getDistrib() {
    	DiscreteDistrib distrib = new DiscreteDistrib(dmatrix, maxLength);
    	return distrib;
    }
    
    
    @Override
    public String toString() {
    	return (new DiscreteDistrib(dmatrix, maxLength)).toString();
    }
}

File: synthetic/randomwalkers.py
import numpy as np

from synthetic.consts import DEFAULT_MAX_DIST


class RandomWalker:
    def __init__(self, net, node, directed, max_length):
        self.net = net
        self.orig = node
        self.directed = directed
        self.max_length = max_length

        self.targ = None
        self.length = -1
        self.forward = False

        self.steps = [-1] * (max_length + 1)

        self.restart()

    def restart(self):
        self.targ = self.orig
        self.length = 0
        self.forward = bool(np.random.randint(0, 2))

    def step(self):
        if self.directed:
            if self.forward:
                neighbors = self.net.out_neighbors(self.targ)
            else:
                neighbors = self.net.in_neighbors(self.targ)
        else:
            neighbors = self.net.neighbors(self.targ)

        if len(neighbors) > 0:
            next_node = neighbors[np.random.randint(0, len(neighbors))]
        else:
            next_node = None

        if next_node is None:
            self.restart()
        elif self.orig == next_node:
            self.restart()
        elif next_node in self.steps:
            self.restart()
        else:
            self.steps[self.length] = next_node
            self.length += 1

            self.targ = next_node

            if self.length > self.max_length:
                self.restart()


class RandomWalkers:
    def __init__(self, net, directed):
        self.net = net
        self.directed = directed

        self.max_length = DEFAULT_MAX_DIST
        self.large_value = self.max_length + 1
        self.steps = 1

        self.nodes = net.graph.vcount()

        self.dmatrix = None
        self.walkers = None

        self.init()
    
    def init(self):
        # clear matrices
        self.dmatrix = np.full((self.nodes, self.nodes), self.large_value)
        for i in range(self.nodes):
            self.dmatrix[i, i] = 0
        
        # init walkers
        self.walkers = [RandomWalker(self.net, node, self.directed, self.max_length) for node in range(self.nodes)]

    def distance(self, x, y):
        return self.dmatrix[x, y]

    def set_distance(self, x, y, d):
        if self.dmatrix[x, y] <= d:
            return

        self.dmatrix[x, y] = d
        
        if not self.directed:
            self.dmatrix[y, x] = d
    
    def step(self):
        for walker in self.walkers:
            if self.net.degree(walker.orig) > 0:
                for _ in range(self.steps):
                    walker.step()
                    if walker.forward:
                        self.set_distance(walker.orig, walker.targ, walker.length)
                    else:
                        self.set_distance(walker.targ, walker.orig, walker.length)
    
    def all_steps(self):
        for _ in range(self.net.graph.ecount()):
            self.step()

    def recompute(self):
        self.init()
        self.all_steps()
